dnssec/main.tf

```diff
@@ -0,0 +1,235 @@
+      ])
+    }
+    "dnssec_errors" = {
+      metric_name = "Errors"
+      desc = join("", [
+        "DNSSEC encountered 1+ errors in <24h",
+        var.dnssec_errors_alarm_desc
+      ])
+    }
+  }
+}
+
+# -- Resources
+
+resource "aws_kms_key" "dnssec" {
+  for_each = var.dnssec_ksks
+  provider = aws.use1
+
+  customer_master_key_spec = "ECC_NIST_P256"
+  deletion_window_in_days  = 7
+  key_usage                = "SIGN_VERIFY"
+  policy                   = data.aws_iam_policy_document.ksk_policy.json
+
+  # These blocks are here, but commented out, because they currently (Dec 2021)
+  # are not configurable with variables / don't support interpolation,
+  # which hinders key rotation significantly.
+  # 
+  # See the README for more info, as well as:
+  # https://github.com/hashicorp/terraform/issues/3116
+  # https://github.com/hashicorp/terraform/issues/4149
+
+  #lifecycle {
+  #  prevent_destroy = true
+  #}
+}
+
+resource "aws_kms_alias" "dnssec" {
+  for_each = var.dnssec_ksks
+  provider = aws.use1
+
+  name          = "alias/${replace(var.dnssec_zone_name, "/\\./", "_")}-ksk-${each.key}"
+  target_key_id = aws_kms_key.dnssec[each.key].key_id
+
+  #lifecycle {
+  #  prevent_destroy = true
+  #}
+}
+
+resource "aws_route53_key_signing_key" "dnssec" {
+  for_each = var.dnssec_ksks
+
+  hosted_zone_id             = var.dnssec_zone_id
+  key_management_service_arn = aws_kms_key.dnssec[each.key].arn
+  name                       = "${var.dnssec_zone_name}-ksk-${each.key}"
+
+  #lifecycle {
+  #  prevent_destroy = true
+  #}
+}
+
+resource "aws_route53_hosted_zone_dnssec" "dnssec" {
+  depends_on = [
+    aws_route53_key_signing_key.dnssec
+  ]
+  hosted_zone_id = var.dnssec_zone_id
+
+  #lifecycle {
+  #  prevent_destroy = true
+  #}
+}
+
+data "aws_iam_policy_document" "dnssec_disable_prevent" {
+  count = var.protect_resources ? 1 : 0
+
+  statement {
+    sid    = "HostedZoneAndKSKDisableDeletePrevent"
+    effect = "Deny"
+    actions = [
+      "route53:DeactivateKeySigningKey",
+      "route53:DeleteHostedZone",
+      "route53:DeleteKeySigningKey",
+      "route53:DisableHostedZoneDNSSEC",
+    ]
+    resources = [
+      "arn:aws:route53:::hostedzone/${var.dnssec_zone_id}"
+    ]
+  }
+
+  dynamic "statement" {
+    for_each = var.dnssec_ksks
+
+    content {
+      sid    = "KMSDisableDeletePreventForAlias${statement.key}"
+      effect = "Deny"
+      actions = [
+        "kms:DeleteAlias",
+      ]
+      resources = [
+        aws_kms_alias.dnssec[statement.key].arn
+      ]
+    }
+  }
+
+  dynamic "statement" {
+    for_each = var.dnssec_ksks
+
+    content {
+      sid    = "KMSKeyDisableDeletePreventForKey${statement.key}"
+      effect = "Deny"
+      actions = [
+        "kms:DisableKey",
+        "kms:ScheduleKeyDeletion",
+      ]
+      resources = [
+        aws_kms_alias.dnssec[statement.key].target_key_arn
+      ]
+    }
+  }
+}
+
+resource "aws_iam_policy" "dnssec_disable_prevent" {
```

Comment:
How do you have a special "I can smack around DNSSEC" role if this Deny is set inside a module?   Can we move the guardrail out to `all` like our RDS delete prevention guardrail?  If we need to ref resources we can do it with `data`, and only the handful of things we care about.  (login.gov, identitysandbox.gov)

Reply:
The IAM policy itself can be created on an optional basis (using the `protect_resources` flag), so its creation can be skipped entirely.

Additionally, this is a policy that gets attached to one or more roles -- meaning that we can select which roles do use it vs. which ones do not. That's how a "break glass" role would work in this case -- you simply create a role which doesn't have this policy attached.